planner/core/logical_plan_builder.go

```diff
@@ -2048,11 +2048,33 @@ func (b *PlanBuilder) buildApplyWithJoinType(outerPlan, innerPlan LogicalPlan, t
 	return ap
 }
 
+func cannotDecorrelate(fromNotIn bool, outerPlan, innerPlan LogicalPlan, condition []expression.Expression) bool {
+	if !fromNotIn {
+		return false
+	}
+	outerSchema := outerPlan.Schema()
+	innerSchema := innerPlan.Schema()
+	cols := make([]*expression.Column, 0, 2*len(condition))
+	cols = expression.ExtractColumnsFromExpressions(cols, condition, nil)
+	for _, col := range cols {
+		if innerCol := innerSchema.RetrieveColumn(col); innerCol != nil && !mysql.HasNotNullFlag(innerCol.RetType.Flag) {
+			return true
+		}
+		if outerCol := outerSchema.RetrieveColumn(col); outerCol != nil && !mysql.HasNotNullFlag(outerCol.RetType.Flag) {
+			return true
+		}
+	}
+	return false
+}
+
 // buildSemiApply builds apply plan with outerPlan and innerPlan, which apply semi-join for every row from outerPlan and the whole innerPlan.
-func (b *PlanBuilder) buildSemiApply(outerPlan, innerPlan LogicalPlan, condition []expression.Expression, asScalar, not bool) (LogicalPlan, error) {
+func (b *PlanBuilder) buildSemiApply(outerPlan, innerPlan LogicalPlan, condition []expression.Expression,
+	asScalar, not bool, fromNotIn bool) (LogicalPlan, error) {
 	b.optFlag = b.optFlag | flagPredicatePushDown
 	b.optFlag = b.optFlag | flagBuildKeyInfo
-	b.optFlag = b.optFlag | flagDecorrelate
+	if !cannotDecorrelate(fromNotIn, outerPlan, innerPlan, condition) {
```

Comment:
This way is not very stable enough.
One query may contain both sub plans can be decorrelated and cannot.

Reply:
Oh you are right, I will mark this `LogicalApply` and add a check to prevent it from being decorrelated when applying `flagDecorrelate`.